Investigate intermittent test failures...again

More intermittent build failures:

```
Error:  Tests run: 12, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 5.593 s <<< FAILURE! - in org.kiwiproject.curator.leader.ManagedLeaderLatchCreatorTest
Error:  org.kiwiproject.curator.leader.ManagedLeaderLatchCreatorTest.shouldRegisterResources_ByDefault  Time elapsed: 0.026 s  <<< ERROR!
org.kiwiproject.curator.leader.exception.ManagedLeaderLatchException: Error starting leader latch
	at org.kiwiproject.curator.leader.ManagedLeaderLatchCreatorTest.shouldRegisterResources_ByDefault(ManagedLeaderLatchCreatorTest.java:228)
Caused by: org.kiwiproject.curator.leader.exception.ManagedLeaderLatchException: Unable to create latch path: /kiwi/leader-latch/test-service/leader-latch
	at org.kiwiproject.curator.leader.ManagedLeaderLatchCreatorTest.shouldRegisterResources_ByDefault(ManagedLeaderLatchCreatorTest.java:228)
Caused by: org.apache.zookeeper.KeeperException$NodeExistsException: KeeperErrorCode = NodeExists for /kiwi/leader-latch/test-service/leader-latch
	at org.kiwiproject.curator.leader.ManagedLeaderLatchCreatorTest.shouldRegisterResources_ByDefault(ManagedLeaderLatchCreatorTest.java:228)

Error:  org.kiwiproject.curator.leader.ManagedLeaderLatchCreatorTest.shouldRegisterListeners_InOrder_UsingVarargs  Time elapsed: 0.015 s  <<< ERROR!
org.kiwiproject.curator.leader.exception.ManagedLeaderLatchException: Error starting leader latch
	at org.kiwiproject.curator.leader.ManagedLeaderLatchCreatorTest.shouldRegisterListeners_InOrder_UsingVarargs(ManagedLeaderLatchCreatorTest.java:150)
Caused by: org.kiwiproject.curator.leader.exception.ManagedLeaderLatchException: Unable to create latch path: /kiwi/leader-latch/test-service/leader-latch
	at org.kiwiproject.curator.leader.ManagedLeaderLatchCreatorTest.shouldRegisterListeners_InOrder_UsingVarargs(ManagedLeaderLatchCreatorTest.java:150)
Caused by: org.apache.zookeeper.KeeperException$NodeExistsException: KeeperErrorCode = NodeExists for /kiwi/leader-latch/test-service/leader-latch
	at org.kiwiproject.curator.leader.ManagedLeaderLatchCreatorTest.shouldRegisterListeners_InOrder_UsingVarargs(ManagedLeaderLatchCreatorTest.java:150)
```

--- a/src/main/java/org/kiwiproject/curator/leader/ManagedLeaderLatchCreator.java
+++ b/src/main/java/org/kiwiproject/curator/leader/ManagedLeaderLatchCreator.java
@@ -3,6 +3,7 @@ package org.kiwiproject.curator.leader;
 import static com.google.common.base.Preconditions.checkState;
 import static com.google.common.collect.Lists.newArrayList;
 import static java.util.Objects.nonNull;
+import static org.kiwiproject.base.KiwiPreconditions.checkArgumentNotNull;
 import static org.kiwiproject.base.KiwiPreconditions.requireNotNull;
 
 import io.dropwizard.setup.Environment;
@@ -40,6 +41,7 @@ public class ManagedLeaderLatchCreator {
 
     // Initialized via instance start() method so cannot be final (unless we used Kotlin and lateinit)
     private ManagedLeaderLatch leaderLatch;
+    private boolean latchStarted;
     private boolean addHealthCheck;
     private ManagedLeaderLatchHealthCheck healthCheck;
     private boolean addResources;
@@ -210,6 +212,12 @@ public class ManagedLeaderLatchCreator {
     }
 
     private void startLatchOrThrow() {
+        startLatchOrThrow(leaderLatch);
+        latchStarted = true;
+    }
+
+    static void startLatchOrThrow(ManagedLeaderLatch leaderLatch) {
+        checkArgumentNotNull(leaderLatch);
         try {
             leaderLatch.start();
         } catch (Exception e) {
@@ -232,7 +240,18 @@ public class ManagedLeaderLatchCreator {
     }
 
     /**
+     * Has this instance created and started a {@link ManagedLeaderLatch}?
+     *
+     * @return true if this instance has created and started a leader latch
+     */
+    public boolean isLeaderLatchStarted() {
+        return latchStarted;
+    }
+
+    /**
      * Returns the {@link ManagedLeaderLatch} created after {@link #start()} has been called.
+     * <p>
+     * Use {@link #isLeaderLatchStarted()} to ensure the latch has been started to ensure this method will succeed.
      *
      * @return the leader latch
      * @throws IllegalStateException if called but the latch has not been started yet
@@ -266,6 +285,6 @@ public class ManagedLeaderLatchCreator {
     }
 
     private void validateStarted() {
-        checkState(nonNull(leaderLatch), "Leader latch is not started; call start() first");
+        checkState(latchStarted, "Leader latch is not started; call start() first");
     }
 }

--- a/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchCreatorTest.java
+++ b/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchCreatorTest.java
@@ -1,15 +1,19 @@
 package org.kiwiproject.curator.leader;
 
+import static java.util.Objects.nonNull;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.awaitility.Awaitility.await;
 import static org.kiwiproject.collect.KiwiLists.first;
 import static org.kiwiproject.collect.KiwiLists.second;
+import static org.kiwiproject.curator.leader.util.CuratorTestHelpers.closeIfStarted;
 import static org.kiwiproject.curator.leader.util.CuratorTestHelpers.deleteRecursivelyIfExists;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyString;
 import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.ArgumentMatchers.isA;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.verifyNoInteractions;
@@ -31,13 +35,17 @@ import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInfo;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.junit.jupiter.api.extension.RegisterExtension;
+import org.kiwiproject.curator.leader.exception.ManagedLeaderLatchException;
 import org.kiwiproject.curator.leader.health.ManagedLeaderLatchHealthCheck;
 import org.kiwiproject.curator.leader.resource.GotLeaderLatchResource;
 import org.kiwiproject.curator.leader.resource.LeaderResource;
+import org.kiwiproject.curator.leader.util.CuratorTestHelpers;
 import org.kiwiproject.test.curator.CuratorTestingServerExtension;
 import org.kiwiproject.test.dropwizard.mockito.DropwizardMockitoMocks;
+import org.kiwiproject.test.junit.jupiter.WhiteBoxTest;
 import org.mockito.ArgumentCaptor;
 
 import java.util.concurrent.TimeUnit;
@@ -58,8 +66,12 @@ class ManagedLeaderLatchCreatorTest {
     private HealthCheckRegistry healthCheckRegistry;
     private ServiceDescriptor serviceDescriptor;
 
+    private ManagedLeaderLatchCreator latchCreator;
+
     @BeforeEach
-    void setUp() {
+    void setUp(TestInfo testInfo) {
+        LOG.info("Set up for test: {}", testInfo.getDisplayName());
+
         client = ZK_TEST_SERVER.getClient();
         var dropwizardMockitoContext = DropwizardMockitoMocks.mockDropwizard();
         environment = dropwizardMockitoContext.environment();
@@ -75,15 +87,31 @@ class ManagedLeaderLatchCreatorTest {
     }
 
     @AfterEach
-    void tearDown() throws Exception {
-        var rootPath = "/kiwi/leader-latch";
+    void tearDown(TestInfo testInfo) throws Exception {
+        LOG.info("Tear down after test: {}", testInfo.getDisplayName());
 
-        deleteRecursivelyIfExists(client, rootPath);
+        if (nonNull(latchCreator) && latchCreator.isLeaderLatchStarted()) {
+            closeIfStarted(latchCreator.getLeaderLatch());
+        }
+
+        var rootPath = "/kiwi/leader-latch";
+        var deleteResult = deleteRecursivelyIfExists(client, rootPath);
+
+        if (deleteResult.failed()) {
+            LOG.error("Path {} was not deleted!", rootPath);
+
+            var kiwiPath = "/kiwi";
+            var childPaths = client.getChildren().forPath(kiwiPath);
+            LOG.info("Children at {}: {}", kiwiPath, childPaths);
+        } else {
+            LOG.info("Path {} delete result: {}", rootPath, deleteResult);
+        }
     }
 
     @Test
     void shouldThrowIllegalStateExceptions_FromGetMethods_WhenNotStarted(SoftAssertions softly) {
-        var latchCreator = ManagedLeaderLatchCreator.from(client, environment, serviceDescriptor);
+        latchCreator = ManagedLeaderLatchCreator.from(client, environment, serviceDescriptor);
+        assertThat(latchCreator.isLeaderLatchStarted()).isFalse();
 
         softlyAssertIllegalStateExceptionThrownBy(softly, latchCreator::getLeaderLatch);
         softlyAssertIllegalStateExceptionThrownBy(softly, latchCreator::getHealthCheck);
@@ -97,9 +125,10 @@ class ManagedLeaderLatchCreatorTest {
 
     @Test
     void shouldCreateAndManageLeaderLatch(SoftAssertions softly) {
-        var latchCreator = ManagedLeaderLatchCreator
+        latchCreator = ManagedLeaderLatchCreator
                 .from(client, environment, serviceDescriptor)
                 .start();
+        assertThat(latchCreator.isLeaderLatchStarted()).isTrue();
 
         assertThat(latchCreator.getLeaderLatch()).isNotNull();
         softly.assertThat(latchCreator.getListeners()).isEmpty();
@@ -123,9 +152,22 @@ class ManagedLeaderLatchCreatorTest {
                 .isEqualTo(LeaderLatch.State.STARTED);
     }
 
+    @WhiteBoxTest
+    void shouldThrowManagedLeaderLatchException_WhenLatchFailsToStart() throws Exception {
+        var leaderLatch = mock(ManagedLeaderLatch.class);
+
+        var ex = new RuntimeException("oop");
+        doThrow(ex).when(leaderLatch).start();
+
+        assertThatThrownBy(() -> ManagedLeaderLatchCreator.startLatchOrThrow(leaderLatch))
+                .isExactlyInstanceOf(ManagedLeaderLatchException.class)
+                .hasMessage("Error starting leader latch")
+                .hasCause(ex);
+    }
+
     @Test
     void shouldIgnoreMultipleCallsToStart() {
-        var latchCreator = ManagedLeaderLatchCreator.from(client, environment, serviceDescriptor);
+        latchCreator = ManagedLeaderLatchCreator.from(client, environment, serviceDescriptor);
 
         var latch1 = latchCreator.start();
         var latch2 = latchCreator.start();
@@ -145,7 +187,7 @@ class ManagedLeaderLatchCreatorTest {
     @Test
     void shouldRegisterListeners_InOrder_UsingVarargs() {
         var noOpListener = new NoOpListener();
-        var latchCreator = ManagedLeaderLatchCreator
+        latchCreator = ManagedLeaderLatchCreator
                 .from(client, environment, serviceDescriptor, noOpListener)
                 .start();
 
@@ -157,7 +199,7 @@ class ManagedLeaderLatchCreatorTest {
     void shouldRegisterListeners_InOrder_UsingAddMethod() {
         var noOpListener = new NoOpListener();
         var loggingListener = new SimpleLoggingListener();
-        var latchCreator = ManagedLeaderLatchCreator.from(client, environment, serviceDescriptor)
+        latchCreator = ManagedLeaderLatchCreator.from(client, environment, serviceDescriptor)
                 .addLeaderLatchListener(noOpListener)
                 .addLeaderLatchListener(loggingListener)
                 .start();
@@ -171,7 +213,7 @@ class ManagedLeaderLatchCreatorTest {
     void shouldReturnImmutableCopyOfListeners() {
         var noOpListener = new NoOpListener();
         var loggingListener = new SimpleLoggingListener();
-        var latchCreator = ManagedLeaderLatchCreator
+        latchCreator = ManagedLeaderLatchCreator
                 .from(client, environment, serviceDescriptor, noOpListener, loggingListener)
                 .start();
 
@@ -206,7 +248,7 @@ class ManagedLeaderLatchCreatorTest {
 
     @Test
     void shouldRegisterHeathCheck_ByDefault() {
-        var latchCreator = ManagedLeaderLatchCreator.from(client, environment, serviceDescriptor).start();
+        latchCreator = ManagedLeaderLatchCreator.from(client, environment, serviceDescriptor).start();
         assertThat(latchCreator.getHealthCheck()).isNotEmpty();
 
         var healthCheck = latchCreator.getHealthCheck().orElseThrow(IllegalStateException::new);
@@ -215,7 +257,7 @@ class ManagedLeaderLatchCreatorTest {
 
     @Test
     void shouldNotRegisterHeathCheck_IfDisabled() {
-        var latchCreator = ManagedLeaderLatchCreator.from(client, environment, serviceDescriptor)
+        latchCreator = ManagedLeaderLatchCreator.from(client, environment, serviceDescriptor)
                 .withoutHealthCheck()
                 .start();
 
@@ -225,7 +267,7 @@ class ManagedLeaderLatchCreatorTest {
 
     @Test
     void shouldRegisterResources_ByDefault() {
-        ManagedLeaderLatchCreator.from(client, environment, serviceDescriptor).start();
+        latchCreator = ManagedLeaderLatchCreator.from(client, environment, serviceDescriptor).start();
 
         verify(jersey).register(isA(GotLeaderLatchResource.class));
         verify(jersey).register(isA(LeaderResource.class));
@@ -233,7 +275,7 @@ class ManagedLeaderLatchCreatorTest {
 
     @Test
     void shouldNotRegisterResources_IfDisabled() {
-        ManagedLeaderLatchCreator.from(client, environment, serviceDescriptor)
+        latchCreator = ManagedLeaderLatchCreator.from(client, environment, serviceDescriptor)
                 .withoutResources()
                 .start();
 
@@ -243,19 +285,24 @@ class ManagedLeaderLatchCreatorTest {
     @Test
     void shouldStartLeaderLatch_WithAllTheDefaultThings() {
         var listener = new BecameLeaderListener();
-        var leaderLatch = ManagedLeaderLatchCreator.startLeaderLatch(client, environment, serviceDescriptor, listener);
+        ManagedLeaderLatch leaderLatch = null;
+        try {
+            leaderLatch = ManagedLeaderLatchCreator.startLeaderLatch(client, environment, serviceDescriptor, listener);
 
-        verify(healthCheckRegistry).register(eq("leaderLatch"), isA(ManagedLeaderLatchHealthCheck.class));
-        verify(jersey).register(isA(GotLeaderLatchResource.class));
-        verify(jersey).register(isA(LeaderResource.class));
+            verify(healthCheckRegistry).register(eq("leaderLatch"), isA(ManagedLeaderLatchHealthCheck.class));
+            verify(jersey).register(isA(GotLeaderLatchResource.class));
+            verify(jersey).register(isA(LeaderResource.class));
 
-        assertStartedAndBecameLeader(leaderLatch, listener);
+            assertStartedAndBecameLeader(leaderLatch, listener);
+        } finally {
+            CuratorTestHelpers.closeIfStarted(leaderLatch);
+        }
     }
 
     @Test
     void shouldStart_WithAllTheDefaultThings() {
         var listener = new BecameLeaderListener();
-        var latchCreator = ManagedLeaderLatchCreator.start(client, environment, serviceDescriptor, listener);
+        latchCreator = ManagedLeaderLatchCreator.start(client, environment, serviceDescriptor, listener);
 
         assertThat(latchCreator.getHealthCheck()).isPresent();
         assertThat(latchCreator.getListeners()).hasSize(1).hasOnlyElementsOfTypes(BecameLeaderListener.class);

--- a/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchTest.java
+++ b/src/test/java/org/kiwiproject/curator/leader/ManagedLeaderLatchTest.java
@@ -1,11 +1,12 @@
 package org.kiwiproject.curator.leader;
 
-import static java.util.Objects.nonNull;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.awaitility.Awaitility.await;
 import static org.awaitility.Durations.FIVE_SECONDS;
+import static org.kiwiproject.curator.leader.util.CuratorTestHelpers.closeIfStarted;
 import static org.kiwiproject.curator.leader.util.CuratorTestHelpers.deleteRecursivelyIfExists;
+import static org.kiwiproject.curator.leader.util.CuratorTestHelpers.startAndAwait;
 import static org.mockito.Mockito.inOrder;
 import static org.mockito.Mockito.mock;
 
@@ -465,26 +466,5 @@ class ManagedLeaderLatchTest {
 
     private Executor nullExecutorWhichShouldNeverBeUsed() {
         return null;
-    }
-
-    private void startAndAwait(ManagedLeaderLatch... latches) throws Exception {
-        for (ManagedLeaderLatch latch : latches) {
-            startAndAwait(latch);
-        }
-    }
-
-    private void startAndAwait(ManagedLeaderLatch latch) throws Exception {
-        latch.start();
-        await().atMost(FIVE_SECONDS).until(latch::isStarted);
-    }
-
-    private void closeIfStarted(ManagedLeaderLatch latch) {
-        if (nonNull(latch) && latch.isStarted()) {
-            LOG.debug("Closing latch {}", latch);
-            latch.stop();
-            await().atMost(FIVE_SECONDS).until(latch::isClosed);
-        } else {
-            LOG.trace("Latch {} not started; ignoring close request", latch);
-        }
     }
 }

--- a/src/test/java/org/kiwiproject/curator/leader/util/CuratorTestHelpers.java
+++ b/src/test/java/org/kiwiproject/curator/leader/util/CuratorTestHelpers.java
@@ -1,10 +1,13 @@
 package org.kiwiproject.curator.leader.util;
 
 import static java.util.Objects.nonNull;
+import static org.awaitility.Awaitility.await;
+import static org.awaitility.Durations.FIVE_SECONDS;
 
 import lombok.experimental.UtilityClass;
 import lombok.extern.slf4j.Slf4j;
 import org.apache.curator.framework.CuratorFramework;
+import org.kiwiproject.curator.leader.ManagedLeaderLatch;
 import org.kiwiproject.retry.SimpleRetryer;
 import org.slf4j.event.Level;
 
@@ -14,15 +17,47 @@ import java.util.concurrent.TimeUnit;
 @Slf4j
 public class CuratorTestHelpers {
 
-    public static void deleteRecursivelyIfExists(CuratorFramework client, String path) throws Exception {
-        if (pathExists(client, path)) {
-            deleteRecursively(client, path);
+    public static void startAndAwait(ManagedLeaderLatch... latches) throws Exception {
+        for (ManagedLeaderLatch latch : latches) {
+            startAndAwait(latch);
         }
     }
 
-    public static void deleteRecursively(CuratorFramework client, String path) {
+    public static void startAndAwait(ManagedLeaderLatch latch) throws Exception {
+        latch.start();
+        await().atMost(FIVE_SECONDS).until(latch::isStarted);
+    }
+
+    public static void closeIfStarted(ManagedLeaderLatch latch) {
+        if (nonNull(latch) && latch.isStarted()) {
+            LOG.debug("Closing latch {}", latch);
+            latch.stop();
+            await().atMost(FIVE_SECONDS).until(latch::isClosed);
+        } else {
+            LOG.trace("Latch {} not started; ignoring close request", latch);
+        }
+    }
+
+    public enum DeleteResult {
+        SUCCEEDED, FAILED, SKIPPED;
+
+        public boolean failed() {
+            return this == FAILED;
+        }
+    }
+
+    public static DeleteResult deleteRecursivelyIfExists(CuratorFramework client, String path) throws Exception {
+        if (pathExists(client, path)) {
+            return deleteRecursively(client, path);
+        }
+
+        return DeleteResult.SKIPPED;
+    }
+
+    public static DeleteResult deleteRecursively(CuratorFramework client, String path) {
         // In GitHub we have seen various intermittent test failures caused by NodeExistsException.
         // See issue: https://github.com/kiwiproject/dropwizard-leader-latch/issues/36
+        // And also: https://github.com/kiwiproject/dropwizard-leader-latch/issues/69
 
         // The following attempts to delete a path and any children. If the path still exists after
         // attempting to delete it, try again up to a max of 5 attempts.
@@ -36,7 +71,7 @@ public class CuratorTestHelpers {
                 .build();
 
         // NOTE: returning null forces a retry unless max attempts have been reached
-        retryer.tryGetObject(() -> {
+        var deletedPathOptional = retryer.tryGetObject(() -> {
             try {
                 deletePathAndChildren(client, path);
 
@@ -46,6 +81,8 @@ public class CuratorTestHelpers {
                 return null;
             }
         });
+
+        return deletedPathOptional.map(thePath -> DeleteResult.SUCCEEDED).orElse(DeleteResult.FAILED);
     }
 
     private static void deletePathAndChildren(CuratorFramework client, String path) throws Exception {

